File: deeplx-spring/src/main/java/com/gchuyun/deeplx/DeepLClient.java
package com.gchuyun.deeplx;


import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpMethod;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.util.List;
import java.util.Random;
import java.util.function.Function;
import java.util.stream.Collectors;

@Component
public class DeepLClient {

  @Autowired
  ObjectMapper objectMapper;

  Logger LOGGER = LoggerFactory.getLogger(DeepLClient.class);
  public Mono<DeepLDTO.ResponseData> translate(String sourceLang, String targetLang, List<String> translateTexts) {
    if (sourceLang == null || sourceLang.trim().isEmpty()) {
      sourceLang = "AUTO";
    }
    if (targetLang == null || targetLang.trim().isEmpty()) {
      if (translateTexts.stream().anyMatch(text -> text.matches(".*[\u4E00-\u9FA5]+.*"))) {
        targetLang = "EN";
      } else {
        targetLang = "ZH";
      }
    }

    long id = getRandomId();
    DeepLDTO.ReqestData deepLReqData = initData(sourceLang, targetLang);
    deepLReqData.setId(id);
    deepLReqData.getParams().setTexts(translateTexts.stream()
      .map(t -> new DeepLDTO.ReqestData.Text() {{
        text = t;
        requestAlternatives = 3;
      }}).collect(Collectors.toList()));
    deepLReqData.getParams().setTimestamp(
      getICountTimeStamp(translateTexts.stream().mapToLong(t -> getICount(t)).sum()));

    Function<Object, String> writeValueAsString = o -> {
      try {
        return objectMapper.writeValueAsString(o);
      } catch (Exception e) {
        throw new RuntimeException(e);
      }
    };
    String postStr = null;
    try {
      postStr = writeValueAsString.apply(deepLReqData);
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
    if ((id + 5) % 29 == 0 || (id + 3) % 13 == 0) {
      postStr = postStr.replace("\"method\":\"", "\"method\" : \"");
    } else {
      postStr = postStr.replace("\"method\":\"", "\"method\": \"");
    }
    if (LOGGER.isDebugEnabled()) {
      LOGGER.debug("deepl request: " + postStr);
    }

    String finalPostStr = postStr;
    return WebClient.builder()
            .baseUrl("https://www2.deepl.com")
            .defaultHeader("Content-Type", "application/json")
            .defaultHeader("Accept", "*/*")
            .defaultHeader("x-app-os-name", "iOS")
            .defaultHeader("x-app-os-version", "16.3.0")
            .defaultHeader("Accept-Language", "en-US,en;q=0.9")
            .defaultHeader("Accept-Encoding", "gzip, deflate, br")
            .defaultHeader("x-app-device", "iPhone13,2")
            .defaultHeader("User-Agent", "DeepL-iOS/2.6.0 iOS 16.3.0 (iPhone13,2)")
            .defaultHeader("x-app-build", "353933")
            .defaultHeader("x-app-version", "2.6")
            .defaultHeader("Connection", "keep-alive")
            .build()
            .method(HttpMethod.POST)
            .uri("/jsonrpc")
            .bodyValue(postStr)
            .exchangeToMono(response -> {
              if (response.statusCode().value() >= 400) {
                return response.bodyToMono(DeepLDTO.ResponseData.ErrorData.class)
                        .doOnSuccess(errorData -> {
                          LOGGER.warn("deepl request: " + finalPostStr);
                          LOGGER.warn("deepl error: " + writeValueAsString.apply(errorData));

                          errorData.setHttpStatusCode(response.statusCode().value());
                        })
                        .map(o -> (DeepLDTO.ResponseData) o);
              } else {
                return response.bodyToMono(DeepLDTO.ResponseData.ResultData.class)
                        .doOnSuccess(deepLRespData -> {
                          if (LOGGER.isDebugEnabled()) {
                            LOGGER.debug("deepl response: " + writeValueAsString.apply(deepLRespData));
                          }
                        })
                        .map(o -> (DeepLDTO.ResponseData) o);
              }
            });
  }
  public DeepLDTO.ReqestData initData(String srcLang, String tgtLang) {
    return new DeepLDTO.ReqestData() {{
      setJsonrpc("2.0");
      setMethod("LMT_handle_texts");
      setParams(new Params() {{
        setSplitting("newlines");
        setLang(new Lang() {{
          setSourceLangUserSelected(srcLang);
          setTargetLang(tgtLang);
        }});
        setCommonJobParams(new CommonJobParams() {{
          setWasSpoken(false);
          setTranscribeAS("");
        }});
      }});
    }};
  }

  public long getICount(String translateText) {
    return translateText.chars().filter(ch -> ch == 'i').count();
  }

  public long getRandomId() {
    Random rand = new Random(System.currentTimeMillis());
    long num = rand.nextInt(99999) + 8300000;
    return num * 1000;
  }

  public long getICountTimeStamp(long iCount) {
    long timeMillis = System.currentTimeMillis();
    if (iCount != 0) {
      iCount = iCount + 1;
      return timeMillis - timeMillis % iCount + iCount;
    } else {
      return timeMillis;
    }
  }
}

File: deeplx-rust/src/deepl_client.rs
use hyper::{Body, Client, Method, Request};
use hyper_tls::HttpsConnector;
use rand::Rng;
use regex::Regex;
use std::time::SystemTime;
use tokio::io::AsyncWriteExt;

lazy_static! {
    static REGEX_HAN: Regex = Regex::new(r".*[\u4E00-\u9FA5]+.*").unwrap();
    static CLIENT = Client::builder().build::<_, Body>(HttpsConnector::new());
}

pub async fn translate(source_lang: &mut String, target_lang: &mut String, translate_texts: Vec<String>) {
    if source_lang.trim().is_empty() {
        source_lang = "AUTO";
    }
    if target_lang.trim().is_empty() {
        if translate_texts.iter().any(|&text| REGEX_HAN.is_match(text)) {
            target_lang = "EN";
        } else {
            target_lang = "ZH";
        }
    }

    let id = get_random_id();
    let data = deepl_dto::RequestData {
        jsonrpc: "2.0",
        method: "LMT_handle_texts",
        id: id,
        params: vec![Params {
            splitting: "newlines",
            lang: Lang {
                source_lang: source_lang,
                target_lang: target_lang,
            },
            commonJobParams: CommonJobParams {
                wasSpoken: false,
                transcribe_as: "",
            },
            texts: translate_texts.map(|t| Text {
                text: t,
                requestAlternatives: 3,
            }),
            timestamp: get_i_count_timestamp(translate_texts.map(|t| get_i_count(t).sum())),
        }],
    };

    let post_str: &String = serde_json::to_string(&data)?;
    if ((id + 5) % 29 == 0 || (id + 3) % 13 == 0) {
        post_str = post_str.replace("\"method\":\"", "\"method\" : \"");
    } else {
        post_str = post_str.replace("\"method\":\"", "\"method\": \"");
    }
    println!("deepl request: {}", &post_str);

    
    let request = Request::builder()
        .method(Method::POST)
        .uri("https://www2.deepl.com/jsonrpc")
        .header("Content-Type", "application/json")
        .header("Accept", "*/*")
        .header("x-app-os-name", "iOS")
        .header("x-app-os-version", "16.3.0")
        .header("Accept-Language", "en-US,en;q=0.9")
        .header("Accept-Encoding", "gzip, deflate, br")
        .header("x-app-device", "iPhone13,2")
        .header("User-Agent", "DeepL-iOS/2.6.0 iOS 16.3.0 (iPhone13,2)")
        .header("x-app-build", "353933")
        .header("x-app-version", "2.6")
        .header("Connection", "keep-alive")
        .body(post_str.into())
        .unwrap();

    let response = CLIENT.request(request).await.unwrap();
    let status = response.status();
    let body = hyper::body::to_bytes(response.into_body()).await.unwrap();
    println!("deepl response: {}", &body);
    if status.is_client_error() || status.is_server_error() {
        let error_data: deepl_dto::Error = serde_json::from_slice(&body).unwrap();
        println!("deepl request: {}", post_str);
        println!("deepl error: {}", to_string(&error_data).unwrap());
        error_data.set_http_status_code(status.as_u16());
    } else {
        let result_data: deepl_dto::Result = from_slice(&body).unwrap();
    }
}

fn init_data(src_lang: String, tgt_lang: String) -> deepl_dto::RequestData {
    deepl_dto::RequestData {
        jsonrpc: "2.0",
        method: "LMT_handle_texts",
        params: vec![Params {
            splitting: "newlines",
            lang: Lang {
                source_lang: src_lang,
                target_lang: tgt_lang,
            },
            commonJobParams: CommonJobParams {
                wasSpoken: false,
                transcribe_as: "",
            },
        }],
    }
}

fn get_i_count(translate_text: &String) -> i64 {
    translate_text.chars().filter(|ch| ch == 'i').count();
}

fn get_random_id() -> i64 {
    rand::thread_rng().gen_range(8300000..8339999) * 1000;
}

fn get_i_count_timestamp(i_count: i64) -> i64 {
    let time_millis = SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .unwrap()
        .as_millis() as i64;
    if i_count != 0 {
        i_count = i_count + 1;
        return time_millis - time_millis % i_count + i_count;
    } else {
        return time_millis;
    }
}
